src/mscorlib/src/System/Threading/ThreadPool.cs

```diff
@@ -458,33 +464,46 @@ internal bool LocalFindAndPop(IThreadPoolWorkItem callback)
 
         public IThreadPoolWorkItem Dequeue(ThreadPoolWorkQueueThreadLocals tl, ref bool missedSteal)
         {
-            WorkStealingQueue localWsq = tl.workStealingQueue;
             IThreadPoolWorkItem callback;
-
-            if ((callback = localWsq.LocalPop()) == null && // first try the local queue
-                !workItems.TryDequeue(out callback)) // then try the global queue
+            int wsqActiveObserved = ThreadPoolGlobals.wsqActive;
+            if(wsqActiveObserved > 0)
             {
-                // finally try to steal from another thread's local queue
-                WorkStealingQueue[] queues = WorkStealingQueueList.Queues;
-                int c = queues.Length;
-                Debug.Assert(c > 0, "There must at least be a queue for this thread.");
-                int maxIndex = c - 1;
-                int i = tl.random.Next(c);
-                while (c > 0)
+                WorkStealingQueue localWsq = tl.workStealingQueue;
+
+                if ((callback = localWsq.LocalPop()) == null && // first try the local queue
+                    !workItems.TryDequeue(out callback)) // then try the global queue
                 {
-                    i = (i < maxIndex) ? i + 1 : 0;
-                    WorkStealingQueue otherQueue = queues[i];
-                    if (otherQueue != localWsq && otherQueue.CanSteal)
+                    // finally try to steal from another thread's local queue
+                    WorkStealingQueue[] queues = WorkStealingQueueList.Queues;
+                    int c = queues.Length;
+                    Debug.Assert(c > 0, "There must at least be a queue for this thread.");
+                    int maxIndex = c - 1;
+                    int i = tl.random.Next(c);
+                    while (c > 0)
                     {
-                        callback = otherQueue.TrySteal(ref missedSteal);
-                        if (callback != null)
+                        i = (i < maxIndex) ? i + 1 : 0;
+                        WorkStealingQueue otherQueue = queues[i];
+                        if (otherQueue != localWsq && otherQueue.CanSteal)
                         {
-                            break;
+                            callback = otherQueue.TrySteal(ref missedSteal);
+                            if (callback != null)
+                            {
+                                break;
+                            }
                         }
+                        c--;
+                    }
+                    if((callback == null) && !missedSteal)
+                    {
+                        Interlocked.CompareExchange(ref ThreadPoolGlobals.wsqActive, wsqActiveObserved - 1, wsqActiveObserved);
```

Comment:
> C# compiler could reorder the read since it could be unaware of the volatile write in the LocalPush(). (Reordering memory reads around a function call may or may not be legal in C#. I suspect it is not.) Marking the read volatile will certainly prevent this and make it explicit.

The compiler probably has to assume that it can't reorder things across a volatile load/store. TryDequeue has one, so it can't reorder the read beyond that.

> arm64 store release semantics is a half barrier. It only orders earlier loads and stores and later load acquires. Later loads and store to other addresses are allowed to pass. Making the read volatile will, turn it into load acquire semantics which will prevent it ordering before the store release

I see, didn't realize it could reorder loads/stores to other addresses. Probably shouldn't assume that it can't anyway. Making both the read and write volatile sounds good to me.